lib/sorcery/controller/config.rb

```diff
@@ -1,69 +1,112 @@
 module Sorcery
   module Controller
-    module Config
-      class << self
-        attr_accessor :submodules
+    class Config
+      DEFAULTS = {
         # what class to use as the user class.
-        attr_accessor :user_class
+        :user_class                           => nil,
+        :submodules                           => [],
         # what controller action to call for non-authenticated users.
-        attr_accessor :not_authenticated_action
+        :not_authenticated_action             => :not_authenticated,
+        :login_sources                        => Set.new,
+        :after_login                          => Set.new,
+        :after_failed_login                   => Set.new,
+        :before_logout                        => Set.new,
+        :after_logout                         => Set.new,
+        :after_remember_me                    => Set.new,
         # when a non logged in user tries to enter a page that requires login,
         # save the URL he wanted to reach, and send him there after login.
-        attr_accessor :save_return_to_url
+        :save_return_to_url                   => true,
         # set domain option for cookies
-        attr_accessor :cookie_domain
+        :cookie_domain                        => nil,
+        :login_session_key                    => :user_id
+      }.freeze
+      private_constant :DEFAULTS
 
-        attr_accessor :login_sources
-        attr_accessor :after_login
-        attr_accessor :after_failed_login
-        attr_accessor :before_logout
-        attr_accessor :after_logout
-        attr_accessor :after_remember_me
+      class << self
+        def instance
+          @instance ||= new(DEFAULTS)
+        end
 
         def init!
-          @defaults = {
-            :@user_class                           => nil,
-            :@submodules                           => [],
-            :@not_authenticated_action             => :not_authenticated,
-            :@login_sources                        => Set.new,
-            :@after_login                          => Set.new,
-            :@after_failed_login                   => Set.new,
-            :@before_logout                        => Set.new,
-            :@after_logout                         => Set.new,
-            :@after_remember_me                    => Set.new,
-            :@save_return_to_url                   => true,
-            :@cookie_domain                        => nil
-          }
+          @instance = new(DEFAULTS)
         end
 
-        # Resets all configuration options to their default values.
-        def reset!
-          @defaults.each do |k, v|
-            instance_variable_set(k, v)
-          end
-        end
+        def add_defaults(defaults)
+          attr_accessor(*defaults.keys)
+
+          # Delegate accessor methods to instance
+          defaults.each_key do |name|
+            class_eval <<-RUBY, __FILE__, __LINE__ + 1
```

Comment:
@sinsoku How does using a multiline string with `, __FILE__, __LINE__ + 1` differ from using a do/end block for class_eval? I'm assuming there's a reason behind doing it this way, but I'm not familiar enough to figure out what it is.

> Evaluates the string or block in the context of mod, except that when a block is given, constant/class variable lookup is not affected

Guessing this might be related, but still haven't grokked it yet.

Reply:
@athix I wrote it according to the following code.
https://github.com/Sorcery/sorcery/blob/a02c1247642357129ea739354c22978a06fccaa9/lib/sorcery/controller/submodules/external.rb#L40-L44

I was interested in the differences, so I investigated and found that string performed better than block.

```ruby
require 'benchmark/ips'

class A
  class << self
    def instance
      "dummy"
    end

    class_eval do
      define_method("with_block") { instance }
    end

    class_eval <<-RUBY, __FILE__, __LINE__ + 1
      def #{"with_str"}
        instance
      end
    RUBY
  end
end

Benchmark.ips do |x|
  x.report("with_block") { A.with_block }
  x.report("with_str") { A.with_str }

  x.compare!
end
```

```
Warming up --------------------------------------
          with_block   714.383k i/100ms
            with_str     1.005M i/100ms
Calculating -------------------------------------
          with_block      7.705M (± 2.8%) i/s -     38.577M in   5.011016s
            with_str     10.466M (± 2.1%) i/s -     53.241M in   5.089350s

Comparison:
            with_str: 10466030.0 i/s
          with_block:  7704795.0 i/s - 1.36x  (± 0.00) slower
```

Reply:
Ahh, so it's a performance improvement. Interesting :slightly_smiling_face: